### PR TITLE
Update Dockerfile

### PR DESCRIPTION
**Pull request by ras07**
Create a non-root user, and run the container as that user.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -70,6 +72,13 @@ COPY run .
 # Allow writing symlinks to build dir
 RUN chown 102:102 app/static/build
 
+# Create user/group to run as
+RUN adduser -D -g $DOCKER_USERID -u $DOCKER_USERID $DOCKER_USER 
+# Fix ownership / permissions
+RUN	chown -R ${DOCKER_USER}:${DOCKER_USER} /whoogle /var/lib/tor && \
```
Comment (benbusby) [inline Dockerfile:78]:
```suggestion
RUN	chown -R ${DOCKER_USER}:${DOCKER_USER} /whoogle /var/lib/tor
```

**Review by benbusby — CHANGES_REQUESTED** [review]
One small issue, otherwise looks good. Thanks!

**Review by benbusby — APPROVED** [review]

**Comment by sethforprivacy** [comment]
I believe that this breaks existing deployments, as the latest image built after this was merged brought my instance down:

```
whoogle-search        | Traceback (most recent call last):
whoogle-search        |   File "/usr/local/lib/python3.8/runpy.py", line 185, in _run_module_as_main
whoogle-search        |     mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
whoogle-search        |   File "/usr/local/lib/python3.8/runpy.py", line 144, in _get_module_details
whoogle-search        |     return _get_module_details(pkg_main_name, error)
whoogle-search        |   File "/usr/local/lib/python3.8/runpy.py", line 111, in _get_module_details
whoogle-search        |     __import__(pkg_name)
whoogle-search        |   File "/whoogle/app/__init__.py", line 116, in <module>
whoogle-search        |     os.symlink(full_cb_path, build_path)
whoogle-search        | PermissionError: [Errno 13] Permission denied: '/whoogle/app/static/css/header.css' -> '/whoogle/app/static/build/header.3f278aa9.css'
whoogle-search        | whoami: unknown uid 102
whoogle-search        | Jan 22 05:16:16.565 [notice] Tor 0.4.6.9 running on Linux with Libevent 2.1.12-stable, OpenSSL 1.1.1l, Zlib 1.2.11, Liblzma 5.2.5, Libzstd 1.5.0 and Unknown N/A as libc.
whoogle-search        | Jan 22 05:16:16.565 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
whoogle-search        | Jan 22 05:16:16.565 [notice] Read configuration file "/etc/tor/torrc".
whoogle-search        | Jan 22 05:16:16.568 [notice] Opening Socks listener on 127.0.0.1:9050
whoogle-search        | Jan 22 05:16:16.568 [notice] Opened Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 22 05:16:16.568 [notice] Opening Control listener on 127.0.0.1:9051
whoogle-search        | Jan 22 05:16:16.568 [notice] Opened Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 22 05:16:16.568 [warn] /var/lib/tor is not owned by this group (root, 0) but by group <unknown> (102).  Are you running Tor as the wrong user?
whoogle-search        | Jan 22 05:16:16.568 [notice] Closing partially-constructed Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 22 05:16:16.568 [notice] Closing partially-constructed Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 22 05:16:16.568 [warn] Failed to parse/validate config: Couldn't create private data directory "/var/lib/tor"
whoogle-search        | Jan 22 05:16:16.568 [err] Reading config failed--see warnings above.
```

Note the "permission denied" error and inability for Tor to create a directory. Does something need to be done by users for this migration to work properly?

**Comment by sethforprivacy** [comment]
I rolled back to the 0.7.0 image without issues, FWIW.

**Comment by ras07** [comment]
Are you running in Docker? If so, can you post your docker-compose file (or docker run command line)?

**Comment by sethforprivacy** [comment]
I changed the user as mentioned in the new docker-compose.yml example, and whoogle now loads but Tor functionality is broken:

```bash
sethforprivacy in apps in ~ via  v14.18.3 took 12s ❯ dco logs --follow whoogle-search
Attaching to whoogle-search
whoogle-search        | Jan 26 19:34:41.504 [notice] Tor 0.4.6.9 running on Linux with Libevent 2.1.12-stable, OpenSSL 1.1.1l, Zlib 1.2.11, Liblzma 5.2.5, Libzstd 1.5.0 and Unknown N/A as libc.
whoogle-search        | Jan 26 19:34:41.504 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
whoogle-search        | Jan 26 19:34:41.505 [notice] Read configuration file "/etc/tor/torrc".
whoogle-search        | Jan 26 19:34:41.508 [notice] Opening Socks listener on 127.0.0.1:9050
whoogle-search        | Jan 26 19:34:41.508 [notice] Opened Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 26 19:34:41.508 [notice] Opening Control listener on 127.0.0.1:9051
whoogle-search        | Jan 26 19:34:41.508 [notice] Opened Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 26 19:34:41.508 [warn] /var/lib/tor is not owned by this user (whoogle, 927) but by <unknown> (102). Perhaps you are running Tor as the wrong user?
whoogle-search        | Jan 26 19:34:41.508 [notice] Closing partially-constructed Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 26 19:34:41.508 [notice] Closing partially-constructed Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 26 19:34:41.508 [warn] Failed to parse/validate config: Couldn't create private data directory "/var/lib/tor"
whoogle-search        | Jan 26 19:34:41.508 [err] Reading config failed--see warnings above.
whoogle-search        | Serving on http://0.0.0.0:5000
```

I would suspect this will be broken for all existing instances of whoogle as the user is being changed but no permission changes are applied.

Might need to add in a permission change or note for this release, @benbusby @ras07.

I'm exploring solutions/workarounds ATM on my own instance now.

**Comment by sethforprivacy** [comment]
Completely removing the docker container and volume and rebuilding does not resolve the issue, either:

```bash
sethforprivacy in apps in ~ via  v14.18.3 ❯ docker rm --force whoogle-search
whoogle-search
sethforprivacy in apps in ~ via  v14.18.3 ❯ docker volume rm d946ac7f5baeb23bf0f2c6c612fe777eae5d5389fb9502cc6e3df088bec1f3f0
d946ac7f5baeb23bf0f2c6c612fe777eae5d5389fb9502cc6e3df088bec1f3f0
sethforprivacy in apps in ~ via  v14.18.3 ❯ dco up -d
Creating whoogle-search ...
Creating whoogle-search ... done
sethforprivacy in apps in ~ via  v14.18.3 took 2s ❯ dco logs --follow whoogle-search
Attaching to whoogle-search
whoogle-search        | Jan 26 19:42:29.636 [notice] Tor 0.4.6.9 running on Linux with Libevent 2.1.12-stable, OpenSSL 1.1.1l, Zlib 1.2.11, Liblzma 5.2.5, Libzstd 1.5.0 and Unknown N/A as libc.
whoogle-search        | Jan 26 19:42:29.636 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
whoogle-search        | Jan 26 19:42:29.636 [notice] Read configuration file "/etc/tor/torrc".
whoogle-search        | Jan 26 19:42:29.640 [notice] Opening Socks listener on 127.0.0.1:9050
whoogle-search        | Jan 26 19:42:29.640 [notice] Opened Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 26 19:42:29.640 [notice] Opening Control listener on 127.0.0.1:9051
whoogle-search        | Jan 26 19:42:29.640 [notice] Opened Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 26 19:42:29.640 [warn] /var/lib/tor is not owned by this user (whoogle, 927) but by <unknown> (102). Perhaps you are running Tor as the wrong user?
whoogle-search        | Jan 26 19:42:29.640 [notice] Closing partially-constructed Socks listener connection (ready) on 127.0.0.1:9050
whoogle-search        | Jan 26 19:42:29.640 [notice] Closing partially-constructed Control listener connection (ready) on 127.0.0.1:9051
whoogle-search        | Jan 26 19:42:29.640 [warn] Failed to parse/validate config: Couldn't create private data directory "/var/lib/tor"
whoogle-search        | Jan 26 19:42:29.640 [err] Reading config failed--see warnings above.
whoogle-search        | Serving on http://0.0.0.0:5000
^CERROR: Aborting.
sethforprivacy in apps in ~ via  v14.18.3 took 17s ❯ docker exec -ti whoogle-search bash
bash-5.1$ ls -la ../
total 72
drwxr-xr-x    1 root     root          4096 Jan 26 19:42 .
drwxr-xr-x    1 root     root          4096 Jan 26 19:42 ..
-rwxr-xr-x    1 root     root             0 Jan 26 19:42 .dockerenv
drwxr-xr-x    1 root     root          4096 Jan 21 20:55 bin
drwxrwxrwx    3 102      102             80 Jan 26 19:42 config
drwxr-xr-x    5 root     root           340 Jan 26 19:42 dev
drwxr-xr-x    1 root     root          4096 Jan 26 19:42 etc
drwxr-xr-x    1 root     root          4096 Jan 21 20:57 home
drwxr-xr-x    1 root     root          4096 Jan 21 20:55 lib
drwxr-xr-x    5 root     root          4096 Nov 24 09:20 media
drwxr-xr-x    2 root     root          4096 Nov 24 09:20 mnt
drwxr-xr-x    2 root     root          4096 Nov 24 09:20 opt
dr-xr-xr-x  470 root     root             0 Jan 26 19:42 proc
drwx------    1 root     root          4096 Nov 30 03:36 root
drwxr-xr-x    1 root     root          4096 Jan 26 19:42 run
drwxr-xr-x    1 root     root          4096 Jan 21 20:55 sbin
drwxr-xr-x    2 root     root          4096 Nov 24 09:20 srv
dr-xr-xr-x   13 root     root             0 Jan 26 19:42 sys
drwxrwxrwt    1 root     root          4096 Nov 30 03:41 tmp
drwxr-xr-x    1 root     root          4096 Jan 21 20:55 usr
drwxr-xr-x    1 root     root          4096 Nov 30 03:40 var
drwxr-xr-x    1 whoogle  whoogle       4096 Jan 21 20:57 whoogle
bash-5.1$ ls -la /var/lib/tor/
total 4
drwxr-xr-x    2 102      102             40 Jan 26 19:42 .
drwxr-xr-x    1 root     root          4096 Jan 21 20:55 .
```

**Comment by sethforprivacy** [comment]
OK, so if I replace the tmpfs lines from the update docker-compose.yml the image works properly.

Will you please add a note that this is a breaking change that requires the new lines (14, 20-22) to be changed before updating to the latest tag or 0.7.1 tag (when released)?

**Comment by benbusby** [comment]
@sethforprivacy what did you have to change those lines to? The tmpfs lines were changed to match the new user ID of the `whoogle` user, so as long as the latest version of both docker-compose.yml and the `latest` tag are pulled, there shouldn't be a need for any manual modification. I'm also curious why tor stopped working, since `/var/lib/tor` should be owned by the `whoogle` user after these changes. At the very least, rebuilding the image from scratch should've produced the correct permissions. 

I'm happy to add/update release notes with whatever is most helpful, I just want to make sure I understand the problem correctly before I recommend anyone goes and changes something when I'm not able to reproduce the issue myself.

**Comment by sethforprivacy** [comment]
Anyone who isn't using the repo won't get any of those changes -- I.e. I copied and pasted the contents of the example compose file way back when I first started hosting Whoogle and its part of a much larger Compose file now. As its not mentioned at all as a breaking change, people won't even know that they need to change the Docker config.

Would be good to call out the changes necessary for running whoogle in the release notes, I.e. "Be sure to update the user: and tmpfs: declarations in your Docked Compose file to match the new user information!" or something along those lines.

**Comment by sethforprivacy** [comment]
To clarify: if anyone had previously pulled the Whoogle service snippet from this repo and put it somewhere else, using the "latest" tag, their Whoogle instance will be completely broken whenever they pull down 0.7.1+ unless they also manually update their Docker Compose file.

**Comment by benbusby** [comment]
Ah, I see. As someone who doesn't personally use docker-compose, it's helpful to know this kind of stuff. I always figured that anyone who opted to use docker-compose would have something in place to ensure upstream changes are pulled in from the repo periodically.

I appreciate the clarification and will update the release notes. I sincerely apologize for the hiccup caused by these changes!

**Comment by benbusby** [comment]
I updated the release notes, let me know if you think the new warning is sufficient to help prevent this from happening to anyone else. Thanks again!

**Comment by sethforprivacy** [comment]
> I updated the release notes, let me know if you think the new warning is sufficient to help prevent this from happening to anyone else. Thanks again!

That looks perfect! Thanks so much for adding that 😄

**Comment by sethforprivacy** [comment]
> Ah, I see. As someone who doesn't personally use docker-compose, it's helpful to know this kind of stuff. I always figured that anyone who opted to use docker-compose would have something in place to ensure upstream changes are pulled in from the repo periodically.
> 
> I appreciate the clarification and will update the release notes. I sincerely apologize for the hiccup caused by these changes!

No apology necessary! Just a part of unintended consequences of well meaning PRs.

Thanks for all of the hard work on this, absolutely amazing tool and one that I use daily.

**Pull request merged**